### PR TITLE
feat: Add combobox system

### PR DESCRIPTION
**Pull request by NicholasBoll**
## Summary

Implements: #1955

This change implements a combobox system and an Autocomplete example using the combobox system.

## Release Category
Components

---

## Checklist

- [x] MDX documentation adheres to Canvas Kit's [standard MDX template](https://github.com/Workday/canvas-kit/discussions/1131)
- [x] Label `ready for review` has been added to PR

## For the Reviewer

- [x] PR title is short and descriptive
- [x] PR summary describes the change (Fixes/Resolves linked correctly)

## Areas for Feedback? (optional)

- [x] Code
- [x] Documentation
- [x] Testing

## Testing Manually

Autocomplete story is found here: https://5d854c26ba934e0020f5e98a-efuqiiwwxd.chromatic.com/?path=/story/components-inputs-combobox--autocomplete

### CODE REVIEW

File: modules/react/combobox/lib/ComboboxCard.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import {Menu} from '@workday/canvas-kit-react/menu';
+
+import {useComboboxModel} from './useComboboxModel';
+
+export interface ComboboxCardProps<T = any> extends ExtractProps<typeof Menu.Card> {}
+
+export const useComboboxCard = createElemPropsHook(useComboboxModel)(model => {
+  return {
+    minWidth: model.state.width,
+  } as const;
```
Comment (mannycarrera4) [inline modules/react/combobox/lib/ComboboxCard.tsx:17]: why is the `as const` needed here?

File: modules/react/combobox/lib/ComboboxCard.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+});
+
+export const ComboboxCard = createSubcomponent('div')({
+  modelHook: useComboboxModel,
+  elemPropsHook: useComboboxCard,
+})<ComboboxCardProps>(({children, ...elemProps}, Element) => {
+  return (
+    <Menu.Card as={Element} {...elemProps}>
```
Comment (mannycarrera4) [inline modules/react/combobox/lib/ComboboxCard.tsx:25]: I know the preference here is to use an elemPropsHook, but you could also pass `model` on line 23, and then you can just add `<Menu.Card as={Element} minWidth={model.state.width} {...elemProps}>`. right? Is the elemPropsHook jut cleaner?

File: modules/react/combobox/lib/ComboboxMenuList.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import {ListBox, ListProps} from '@workday/canvas-kit-react/collection';
+
+import {useComboboxModel} from './useComboboxModel';
+
+export interface ComboboxMenuListProps<T = any>
+  extends Omit<ExtractProps<typeof ListBox>, 'children'> {
+  /**
+   * Children of the Combobox. Should contain a `<Combobox.Target>`, a `<Combobox.Content>`
+   */
+  children: ListProps<T>['children'];
```
Comment (mannycarrera4) [inline modules/react/combobox/lib/ComboboxMenuList.tsx:19]: we use this type quite frequently: `React.ReactNode | ((item: T) => React.ReactNode)` or `ListProps['children']`. I wonder if it's worth adding that to our lower level components, or use `ListProps['children']` where ever it expects a list of items like action bar list, or tabs. Sometimes we define the type, sometimes we use listProps

File: modules/react/combobox/lib/ComboboxCard.tsx (hunk 1 above)
Comment (NicholasBoll) [reply modules/react/combobox/lib/ComboboxCard.tsx:17]: It is not needed here for this case, but if the attribute was `position`, there would be a Typescript error saying something like "string is not assignable to `'relative' | 'absolute' | 'static' | ...`. The `as const` makes the object be readonly which narrows types. In this case, there's nothing to narrow.

File: modules/react/combobox/lib/ComboboxCard.tsx (hunk 2 above)
Comment (NicholasBoll) [reply modules/react/combobox/lib/ComboboxCard.tsx:25]: The preference is to only access the model in elemProps hooks. It is available to the component, but should not be the default.

File: modules/react/combobox/lib/ComboboxMenuList.tsx (hunk 3 above)
Comment (NicholasBoll) [reply modules/react/combobox/lib/ComboboxMenuList.tsx:19]: I've been converting to use `ListProps['children']` instead. I was considering having these interfaces extend from `ListProps`. The problem is we tend to add custom JSDocs to the `children` prop. The actual value might only matter for doc parsing.

File: modules/react/collection/lib/useListActiveDescendant.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import {createElemPropsHook, slugify} from '@workday/canvas-kit-react/common';
+
+import {useCursorListModel} from './useCursorListModel';
+
+/**
+ * This elemProps hook is used for cursor navigation by using [Active
+ * Descendant](https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#kbd_focus_activedescendant).
+ * `aria-activedescendant` tells assistive technology which element has "focus" within the component
+ * while this single element maintains focus.
```
Comment (mannycarrera4) [inline modules/react/collection/lib/useListActiveDescendant.ts:10]: ahh, is this to indicate the input to keep focus while there are focusable elements as descendant

File: modules/react/collection/lib/ListBox.tsx [hunk 5]
```diff
@@ -64,20 +64,24 @@ export const ListBox = createContainer('ul')({
...
       ref={model.state.containerRef}
       height={
         height || model.state.isVirtualized ? model.state.UNSTABLE_virtual.totalSize : undefined
       }
+      marginY={marginY}
       maxHeight={maxHeight}
       overflowY={model.state.orientation === 'vertical' ? 'auto' : undefined}
     >
-      <Box as={Element} {...elemProps}>
+      <Flex as={Element} {...elemProps} marginY={0}>
```
Comment (mannycarrera4) [inline modules/react/collection/lib/ListBox.tsx:82]: is there ever a reason consumers MIGHt want to override this?

File: modules/labs-react/combobox/lib/Combobox.tsx [hunk 6]
```diff
@@ -319,7 +319,7 @@ export const Combobox = ({
       }
     }
 
-    setIsOpened(false);
+    // setIsOpened(false);
```
Comment (willklein) [inline modules/labs-react/combobox/lib/Combobox.tsx:322]: Leftover commented line

File: modules/react/collection/lib/ListBox.tsx (hunk 5 above)
Comment (NicholasBoll) [reply modules/react/collection/lib/ListBox.tsx:82]: It cannot be overridden. It _will_ mess up the proper rendering of a virtual list

Comment (NicholasBoll) [reply modules/react/collection/lib/ListBox.tsx:82]: You can see that `marginY` is extraced and used above.

File: cypress/integration/Autocomplete.spec.ts [hunk 7]
```diff
@@ -0,0 +1,348 @@
+import * as h from '../helpers';
```
Comment (NicholasBoll) [inline cypress/integration/Autocomplete.spec.ts:1]: This whole file is copied from https://github.com/Workday/canvas-kit/blob/9304d136509aed73b0f272dc62d4609d8b91be36/cypress/integration/ComboboxLabs.spec.ts

Some lines were updated to point to the correct story and some tests are skipped because grouping isn't supported yet.

File: modules/react/common/lib/utils/components.ts [hunk 8]
```diff
@@ -663,13 +661,14 @@ export const createHook = <M extends Model<any, any>, PO extends {}, PI extends
 };
 
 /**
+ * @deprecated use `createSubModelElemPropsHook` instead
```
Comment (alanbsmith) [inline modules/react/common/lib/utils/components.ts:664]:
```suggestion
```

Comment (alanbsmith) [reply modules/react/common/lib/utils/components.ts:664]: Looks like a dupe.

Comment (NicholasBoll) [reply modules/react/common/lib/utils/components.ts:664]: Merge dupe

File: modules/react/collection/lib/ListBox.tsx [hunk 9]
```diff
@@ -4,21 +4,21 @@ import {
...
+export interface ListBoxProps<T = any> extends Omit<FlexProps, 'children'> {
+  children?: React.ReactNode | ((item: T, index: number) => React.ReactNode);
 }
 
 export const ListBoxItem = createSubcomponent('li')({
   displayName: 'Item',
   modelHook: useListModel,
   elemPropsHook: useListItemRegister,
-})<BoxProps>((elemProps, Element) => {
+})<FlexProps>((elemProps, Element) => {
```
Comment (mannycarrera4) [inline modules/react/collection/lib/ListBox.tsx:21]:
```suggestion
})<ExtractProps<typeof Flex>>((elemProps, Element) => {
```

File: modules/react/collection/lib/ListBox.tsx [hunk 10]
```diff
@@ -4,21 +4,21 @@ import {
   createElemPropsHook,
   createSubcomponent,
```
Comment (mannycarrera4) [inline modules/react/collection/lib/ListBox.tsx:5]:
```suggestion
  createSubcomponent,
  ExtractProps
```

File: modules/react/collection/lib/ListBox.tsx (hunk 9 above)
Comment (NicholasBoll) [reply modules/react/collection/lib/ListBox.tsx:21]: You forget the `never`:
```suggestion
})<ExtractProps<typeof Flex, never>>((elemProps, Element) => {
```

File: modules/react/collection/lib/ListBox.tsx [hunk 11]
```diff
@@ -64,20 +64,24 @@ export const ListBox = createContainer('ul')({
...
-})<ListBoxProps>(({height, maxHeight, ...elemProps}, Element, model) => {
+})<ListBoxProps>(({height, maxHeight, marginY, ...elemProps}, Element, model) => {
+  // We're moving `marginY` to the container to not interfere with the virtualization size. We set
+  // the `marginY` on the Flex to `0` to avoid inaccurate scrollbars
+
   // TODO figure out what style props should go to which `Box`
   return (
     <Box
       ref={model.state.containerRef}
       height={
```
Comment (mannycarrera4) [inline modules/react/collection/lib/ListBox.tsx:75]: not part of this PR, but do we want to put this in a `styled` attribute?

File: modules/react/collection/lib/ListBox.tsx [hunk 12]
```diff
@@ -64,20 +64,24 @@ export const ListBox = createContainer('ul')({
...
         height || model.state.isVirtualized ? model.state.UNSTABLE_virtual.totalSize : undefined
```
Comment (mannycarrera4) [inline modules/react/collection/lib/ListBox.tsx:76]: not part of this PR, but do we want to put this in a `styled` attribute?

File: modules/react/collection/lib/useListLoader.ts [hunk 13]
```diff
@@ -13,9 +15,19 @@ export interface LoadParams {
   filter: CollectionLoader['filter'];
   sorter: CollectionLoader['sorter'];
 }
-export interface AsyncCollectionConfig<T> {
+export interface AsyncCollectionConfig<T, M extends Model<any, any>> {
   pageSize: number;
```
Comment (mannycarrera4) [inline modules/react/collection/lib/useListLoader.ts:19]:
```suggestion
  /**
   * The total number of results to show per page
   */
  pageSize: number;
```

File: modules/react/collection/lib/useOverflowListTarget.tsx [hunk 14]
```diff
@@ -14,33 +14,31 @@ const hiddenStyle = {
...
+    }
+  });
+
+  const isHidden = !model.state.hiddenIds.length;
+
+  return {
+    ref: elementRef,
+    'aria-hidden': isHidden,
+    tabIndex: isHidden ? -1 : 0,
+    style: isHidden ? hiddenStyle : {},
```
Comment (mannycarrera4) [inline modules/react/collection/lib/useOverflowListTarget.tsx:42]:
```suggestion
    style: isHidden ? hiddenStyle : undefined,
```
??? will that create an empty style attribute?

File: modules/react/combobox/README.md [hunk 15]
```diff
@@ -0,0 +1,9 @@
+
+# Canvas Kit Combobox
+
+Combobox
+
+View the [documentation for Combobox](https://workday.github.io/canvas-kit/?path=/docs/combobox-react)
```
Comment (mannycarrera4) [inline modules/react/combobox/README.md:6]:
```suggestion
View the [documentation for Combobox](https://workday.github.io/canvas-kit/?path=/docs/features-combobox--autocomplete)
```

File: modules/react/combobox/lib/Combobox.tsx [hunk 16]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+
+import {createContainer} from '@workday/canvas-kit-react/common';
+
+import {useComboboxModel} from './useComboboxModel';
+import {ComboboxInput} from './ComboboxInput';
+import {ComboboxMenu} from './ComboboxMenu';
+export interface ComboboxProps {
+  /**
+   * Children of the Combobox. Should contain a `<Combobox.Input>`, a `<Combobox.Content>`
```
Comment (mannycarrera4) [inline modules/react/combobox/lib/Combobox.tsx:10]:
```suggestion
   * Children of the Combobox. Should contain a `<Combobox.Input>` and a `<Combobox.Content>`
```

Comment (mannycarrera4) [reply modules/react/combobox/lib/Combobox.tsx:10]: content or menu?

File: modules/react/combobox/lib/ComboboxInput.tsx [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+      ref: elementRef,
+    } as const;
+  }),
+  useComboboxInputOpenWithArrowKeys,
+  useListActiveDescendant,
+  useListKeyboardHandler,
+  usePopupTarget
+);
+
+export const ComboboxInput = createSubcomponent(TextInput)({
```
Comment (mannycarrera4) [inline modules/react/combobox/lib/ComboboxInput.tsx:93]: TextInput doesn't extend box props, I'm sure people will want to customize this. Non blocking, just something to think about when we re work out form inputs

File: modules/react/combobox/lib/ComboboxMenuItem.tsx [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+  useListItemAllowChildStrings,
+  useListItemRegister,
+} from '@workday/canvas-kit-react/collection';
+import {OverflowTooltip} from '@workday/canvas-kit-react/tooltip';
+
+import {useComboboxModel} from './useComboboxModel';
+
+export interface ComboboxMenuItemProps {
+  /**
+   * Children of the Combobox. Should contain a `<Combobox.Target>`, a `<Combobox.Content>`
```
Comment (mannycarrera4) [inline modules/react/combobox/lib/ComboboxMenuItem.tsx:19]: `Combobox.Menu` ?

Comment (mannycarrera4) [inline modules/react/combobox/lib/ComboboxMenuItem.tsx:19]:
```suggestion
   * Children of the Combobox. Should contain a `<Combobox.Target>` and a `<Combobox.Menu>`
```

File: modules/react/combobox/lib/ComboboxMenuItem.tsx [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+  children: React.ReactNode;
+}
+
+export const useComboboxMenuItem = composeHooks(
+  createElemPropsHook(useMenuModel)((model, _, elemProps: {'data-id'?: string} = {}) => {
+    const id = elemProps['data-id'] || '';
+    const onMouseDown = (event: React.MouseEvent<HTMLElement>) => {
+      if (
+        model.state.nonInteractiveIds.includes(id) ||
+        event.currentTarget.hasAttribute('aria-disabled')
```
Comment (mannycarrera4) [inline modules/react/combobox/lib/ComboboxMenuItem.tsx:30]: should this account is something also adds `disabled` ?

File: utils/create-component/createReactComponent.js [hunk 20]
```diff
@@ -35,7 +35,7 @@ module.exports = (modulePath, name, description, prerelease, category) => {
   const pascalCaseName = getPascalCaseName(name);
   const titleCaseName = getTitleCaseName(name);
   const rootPath = '../../..';
-  const storyPath = `${prereleaseTitle || `Components/${category}`}/${titleCaseName}/React`;
+  const storyPath = `${prereleaseTitle || `Components/${category}`}/${titleCaseName}`;
```
Comment (mannycarrera4) [inline utils/create-component/createReactComponent.js:38]: nice

File: modules/react/menu/index.ts [hunk 21]
```diff
@@ -2,7 +2,7 @@ export {useMenuModel} from './lib/useMenuModel';
 export {Menu} from './lib/Menu';
 
 export {useMenuCard} from './lib/MenuCard';
-export {useMenuItem} from './lib/MenuItem';
+export {useMenuItem, StyledMenuItem} from './lib/MenuItem';
```
Comment (mannycarrera4) [inline modules/react/menu/index.ts:5]: does this have to go into the upgrade guide? I know combobox in labs did this

File: modules/react/common/lib/utils/components.ts [hunk 22]
```diff
@@ -592,10 +592,8 @@ export const createElemPropsHook = <TModelHook extends (config: any) => Model<an
   modelHook: TModelHook
 ) => <PO extends {}, PI extends {}>(
   fn: (
-    model: TModelHook extends (config: any) => infer TModel
-      ? TModel
-      : {state: Record<string, any>; events: Record<string, any>},
-    ref?: React.Ref<any>,
+    model: TModelHook extends (config: any) => infer TModel ? TModel : Model<any, any>,
+    ref?: React.Ref<unknown>,
```
Comment (mannycarrera4) [inline modules/react/common/lib/utils/components.ts:596]: does the same need to be added to `createSubComponent`?

File: modules/react/combobox/stories/visual-testing/stories_Combobox.tsx [hunk 23]
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
+
+import {StaticStates, ComponentStatesTable} from '@workday/canvas-kit-react/testing';
+import {withSnapshotsEnabled} from '../../../../../utils/storybook';
+
+import {Combobox} from '@workday/canvas-kit-react/combobox';
+
+export default withSnapshotsEnabled({
+  title: 'Testing/Inputs/Combobox',
```
Comment (mannycarrera4) [inline modules/react/combobox/stories/visual-testing/stories_Combobox.tsx:9]:
```suggestion
  title: 'Testing/Combobox',
```

File: modules/react/collection/lib/ListBox.tsx (hunk 12 above)
Comment (NicholasBoll) [reply modules/react/collection/lib/ListBox.tsx:77]: Possibly.

Comment (NicholasBoll) [reply modules/react/collection/lib/ListBox.tsx:77]: heh. It already is

File: modules/docs/docgen/spec/enhancedComponentParser.spec.ts [hunk 24]
```diff
@@ -33,7 +33,7 @@ describe('enhancedComponentParser', () => {
   });
 
   describe('elemPropsHook', () => {
-    it('should handle "createElemPropsHook" with elemProps defined', () => {
+    it.only('should handle "createElemPropsHook" with elemProps defined', () => {
```
Comment (mannycarrera4) [inline modules/docs/docgen/spec/enhancedComponentParser.spec.ts:36]: remove .only

File: modules/react/combobox/lib/ComboboxMenuList.tsx [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+ * The `listbox` uses `aria-labelledby` pointing to the input of a combobox. This input should be
+ * labelled by a form field label for proper accessibility. Use {@link FormField} to ensure proper
+ * accessibility.
+ */
+export const useComboboxMenuList = createElemPropsHook(useMenuModel)(model => {
+  return {
+    role: 'listbox',
+    'aria-labelledby': model.state.id,
+    id: `${model.state.id}-list`,
+    flexDirection: model.state.orientation === 'vertical' ? 'column' : 'row',
```
Comment (mannycarrera4) [inline modules/react/combobox/lib/ComboboxMenuList.tsx:22]:
```suggestion
    style: { flexDirection: model.state.orientation === 'vertical' ? 'column' : 'row'},
```

Comment (mannycarrera4) [reply modules/react/combobox/lib/ComboboxMenuList.tsx:22]: idk if we also want to do this in elemPropsHook

File: modules/react/collection/lib/useOverflowListTarget.tsx (hunk 14 above)
Comment (NicholasBoll) [reply modules/react/collection/lib/useOverflowListTarget.tsx:42]: React doesn't apply an attribute if the value is `undefined`

File: modules/react/combobox/lib/Combobox.tsx (hunk 16 above)
Comment (NicholasBoll) [reply modules/react/combobox/lib/Combobox.tsx:10]: Menu. This was autogenerated by the `yarn create-component` script

File: modules/react/combobox/lib/ComboboxInput.tsx (hunk 17 above)
Comment (NicholasBoll) [reply modules/react/combobox/lib/ComboboxInput.tsx:99]: We may be able to have `TextInput` extend box in a minor release. It should only add props.

File: modules/react/combobox/lib/ComboboxMenuItem.tsx (hunk 19 above)
Comment (NicholasBoll) [reply modules/react/combobox/lib/ComboboxMenuItem.tsx:30]: `disabled` is not valid on an item and if it is added, no events will be fired on it.

Comment (NicholasBoll) [reply modules/react/combobox/lib/ComboboxMenuItem.tsx:30]: I should clarify:

```jsx
<button disabled onMouseDown={() => console.log('The browser will prevent this event from ever firing')} />
```

File: modules/react/menu/index.ts (hunk 21 above)
Comment (NicholasBoll) [reply modules/react/menu/index.ts:5]: Possibly? It is a new feature...

File: modules/react/combobox/lib/ComboboxMenuList.tsx (hunk 25 above)
Comment (NicholasBoll) [reply modules/react/combobox/lib/ComboboxMenuList.tsx:22]: The direction should not change during the lifecycle of the component. It does not need to be a `style`.

Typically all props that need access to the model should be in the hook. I'm also wondering if we want ALL props to be in the hook.

File: modules/react/combobox/stories/Combobox.stories.mdx [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+# Combobox
+
+Combobox is an abstract compound component. The term "Combobox" is based on the
+[Combobox Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/) as defined in the ARIA
+Authoring Practices Guide (APG):
+
+> A [combobox](https://w3c.github.io/aria/#combobox) is an input widget with an associated popup
+> that enables users to select a value for the combobox from a collection of possible values.
+
+Autocomplete and Select are examples of a Combobox and these components should be used instead.
```
Comment (jamesfan) [inline modules/react/combobox/stories/Combobox.stories.mdx:17]: This implies Canvas Kit has Autocomplete and Select components that use Combobox, which we don't (yet). What do you think about this instead:

```suggestion
Combobox is not a standalone component and must be used in conjunction with other pieces to create
something usable.
```

Open to suggestions. "Pieces" is also intentionally vague, but open to alternatives there as well.

File: modules/react/combobox/stories/Combobox.stories.mdx [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+```sh
+yarn add @workday/canvas-kit-react
+```
+
+## Usage
+
+### Autocomplete
+
+This example shows an Autocomplete example using `FormField`, `InputGroup`, and the `Combobox`
+components to make an autocomplete form field. It uses the `useComboboxLoader` to make mock API
```
Comment (jamesfan) [inline modules/react/combobox/stories/Combobox.stories.mdx:30]:
```suggestion
components to make an autocomplete form field. It uses the `useComboboxLoader` hook to make mock API
```

or

```suggestion
components to make an autocomplete form field. It uses `useComboboxLoader` to make mock API
```

File: modules/react/combobox/README.md [hunk 28]
```diff
@@ -0,0 +1,9 @@
+# Canvas Kit Combobox
+
+Combobox
+
+View the
+[documentation for Combobox](https://workday.github.io/canvas-kit/?path=/docs/features-combobox--autocomplete)
+on Storybook.
+
+[> Workday Design Reference: Combobox](https://design.workday.com/components/)
```
Comment (jamesfan) [inline modules/react/combobox/README.md:9]: No point in this line being here.
```suggestion
```

File: utils/create-component/templates/react/component.js [hunk 29]
```diff
@@ -15,11 +15,20 @@ export interface ${pascalCaseName}Props extends ExtractProps<typeof Box, never>
...
 
+/**
+ * JSDoc for ${pascalCaseName}. Will be part of the Component API docs
+ */
 export const ${pascalCaseName} = createContainer('div')({
   displayName: '${pascalCaseName}',
   modelHook: use${pascalCaseName}Model,
   subComponents: {
+    /**
+     * JSDoc for the ${pascalCaseName}.Target. Will be part of the ComponentAPI docs
```
Comment (jamesfan) [inline utils/create-component/templates/react/component.js:26]:
```suggestion
     * JSDoc for ${pascalCaseName}.Target. Will be part of the Component API docs
```

File: modules/react/combobox/lib/ComboboxMenu.tsx [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+     *     </Combobox>
+     *   )
+     * }
+     * ```
+     */
+    List: ComboboxMenuList,
+    /**
+     * A combobox item has an optional `data-id` prop that identifies the item in the combobox
+     * menu list and will be passed to the optional `onSelect` callback of the combobox model. A
+     * combobox item can contain any HTML.
```
Comment (jamesfan) [inline modules/react/combobox/lib/ComboboxMenu.tsx:54]:
```suggestion
     * `Combobox.Menu.Item` has an optional `data-id` prop that identifies the item in the
     * `Combobox.Menu.List` and will be passed to the optional `onSelect` callback of the
     * `ComboboxModel`. `Combobox.Menu.Item` can contain any HTML.
```

File: modules/react/combobox/lib/ComboboxMenu.tsx [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    Item: ComboboxMenuItem,
+    /**
+     * The combobox card is a non-semantic element used to give the dropdown menu its distinct visual
+     * cue that the dropdown menu is floating above other content. A combobox card usually contains a
+     * menu list, but can also contain other elements like a header or footer.
```
Comment (jamesfan) [inline modules/react/combobox/lib/ComboboxMenu.tsx:60]:
```suggestion
     * `Combobox.Menu.Card` is a non-semantic element used to give the dropdown menu its distinct visual
     * cue that the dropdown menu is floating above other content. `Combobox.Menu.Card` usually contains a
     * `Combobox.Menu.List`, but it can also contain other elements like a header or footer.
```

File: modules/react/combobox/lib/ComboboxMenuList.tsx [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+import {createElemPropsHook, createSubcomponent} from '@workday/canvas-kit-react/common';
+import {useMenuModel} from '@workday/canvas-kit-react/menu';
+import {ListBox, ListBoxProps} from '@workday/canvas-kit-react/collection';
+
+import {useComboboxModel} from './useComboboxModel';
+
+export interface ComboboxMenuListProps<T = any> extends ListBoxProps<T> {}
+
+/**
+ * The `listbox` uses `aria-labelledby` pointing to the input of a combobox. This input should be
```
Comment (jamesfan) [inline modules/react/combobox/lib/ComboboxMenuList.tsx:13]:
```suggestion
 * The `listbox` uses `aria-labelledby` pointing to the `Combobox.Input`. This input should be
```

File: modules/react/combobox/lib/useComboboxInputOpenWithArrowKeys.ts [hunk 33]
```diff
@@ -0,0 +1,21 @@
+import {createElemPropsHook} from '@workday/canvas-kit-react/common';
+
+import {useComboboxModel} from './useComboboxModel';
+
+/**
+ * Opens a combobox with the up/down arrow keys. This should be used only on the input element of a
+ * combobox. It adds a keydown listener for the up/down arrow keys.
```
Comment (jamesfan) [inline modules/react/combobox/lib/useComboboxInputOpenWithArrowKeys.ts:7]:
```suggestion
* Opens a `Combobox` with the up/down arrow keys. This should be used only on a
 * {@link ComboboxInput Combobox.Input}. It adds a keydown listener for the
 * up/down arrow keys.
```

File: modules/react/combobox/lib/useComboboxLoader.ts [hunk 34]
```diff
@@ -0,0 +1,71 @@
...
+  return (fn: (...args: any[]) => any, wait: number) => {
+    clearTimeout(debounceTimer.current);
+
+    debounceTimer.current = (setTimeout(fn, wait) as any) as number; // We don't actually care what the real return type of setTimeout is. It is a number in the browser and something else in Node
+  };
+};
+
+/**
+ * Create a combobox data loader and a model. The combobox loader extends the {@link useListLoader}
+ * and connects a combobox input to the filter of the data loader. A simple loader using
```
Comment (jamesfan) [inline modules/react/combobox/lib/useComboboxLoader.ts:28]:
```suggestion
* Creates a `Combobox` data loader and a model. The `Combobox` loader extends the {@link useListLoader}
 * and connects a {@link ComboboxInput Combobox.Input} to the filter of the data loader. A simple loader using
```

File: modules/react/combobox/lib/useComboboxModel.tsx [hunk 35]
```diff
@@ -0,0 +1,85 @@
...
+
+  return {
+    onChange: config.onChange,
+    state: {value: config.value, inputRef},
+    events: {},
+  };
+});
+
+/**
+ * The ComboboxModel extends the {@link ListModel} and the {@link InputModel}. Selecting items from
```
Comment (jamesfan) [inline modules/react/combobox/lib/useComboboxModel.tsx:31]:
```suggestion
 * `ComboboxModel` extends the {@link ListModel} and the {@link InputModel}. Selecting items from
```

**Review by jamesfan — COMMENTED** [review]
Leaving a number of suggestions here to use the code backtick more frequently when referencing components and subcomponents. We don't do this consistently in our component documentation, but I'd like to start making it a practice.

File: modules/react/combobox/lib/ComboboxMenu.tsx [hunk 36]
```diff
@@ -0,0 +1,67 @@
...
+     *     items: ['First Item', 'Second Item']
+     *   })
+     *
+     *   return (
+     *     <Combobox model={model}>
+     *       // other combobox subcomponents
+     *       <Combobox.List>
+     *         {(item) => <Combobox.Item>{item}</Combobox.Item>}
+     *       </Combobox.List>
+     *     </Combobox>
```
Comment (jamesfan) [inline modules/react/combobox/lib/ComboboxMenu.tsx:45]:
```suggestion
     *     <Combobox model={model}>
     *       // other combobox subcomponents
     *       <Combobox.Menu.List>
     *         {(item) => <Combobox.Menu.Item>{item}</Combobox.Menu.Item>}
     *       </Combobox.Menu.List>
     *     </Combobox>
```

Not sure if we need to wrap the `Combobox.Menu.List` in a `Combobox.Menu.Popper` and `Combobox.Menu.Card` as well or if that's what "other combobox subcomponents" is meant to represent.

Comment (jamesfan) [inline modules/react/combobox/lib/ComboboxMenu.tsx:45]:
```suggestion
     *     <Combobox model={model}>
     *       // other combobox subcomponents
     *       <Combobox.Menu.List>
     *         {(item) => <Combobox.Menu.Item>{item}</Combobox.Menu.Item>}
     *       </Combobox.Menu.List>
     *     </Combobox>
```

Not sure if we need to wrap the `Combobox.Menu.List` in a `Combobox.Menu.Popper` and `Combobox.Menu.Card` as well or if that's what "other combobox subcomponents" is meant to represent.

File: modules/react/combobox/lib/ComboboxMenuList.tsx (hunk 25 above)
Comment (mannycarrera4) [reply modules/react/combobox/lib/ComboboxMenuList.tsx:22]: hmm fair point, idk we've made that distinction, my assumption is, if the prop needs access to the model, create an elemProps hooks, if it doesn't then it's not part of the hook

File: modules/react/breadcrumbs/lib/BreadcrumbsItem.tsx [hunk 37]
```diff
@@ -38,7 +38,7 @@ export const useBreadcrumbsItem = composeHooks(
   createElemPropsHook(useBreadcrumbsModel)(
     (
       {state},
-      _?: React.Ref<HTMLElement>,
+      _,
```
Comment (alanbsmith) [inline modules/react/breadcrumbs/lib/BreadcrumbsItem.tsx:41]: Question: is moving this from optional to required a breaking change? Do we need to help consumers update?

File: modules/react/menu/lib/Menu.tsx [hunk 38]
```diff
@@ -92,8 +92,8 @@ export const Menu = createContainer()({
      */
     List: MenuList,
     /**
-     * A menu item has an optional `name` prop that identifies the item in the menu list and will be
-     * passed to the optional `onSelect` callback of the menu model. A menu item can contain any
+     * A menu item has an optional `data-id` prop that identifies the item in the menu list and will
+     * be passed to the optional `onSelect` callback of the menu model. A menu item can contain any
```
Comment (alanbsmith) [inline modules/react/menu/lib/Menu.tsx:96]:
```suggestion
     * A `Menu.Item` has an optional `data-id` prop that identifies the item in the `Menu.List` and will
     * be passed to the optional `onSelect` callback of the `Menu` model. A `Menu.Item` can contain any
```

File: modules/react/menu/lib/Menu.tsx [hunk 39]
```diff
@@ -92,8 +92,8 @@ export const Menu = createContainer()({
...
      * HTML. If more complex HTML is provided, consider
```
Comment (alanbsmith) [inline modules/react/menu/lib/Menu.tsx:97]: > "If more complex HTML is provided, consider..."

Consider what?

**Review by alanbsmith — APPROVED** [review]
Really nice work, Nicholas. Thank you! Left a few comments and a question. I agree with @jamesfan's suggestions as well. But I don't see anything that should block the PR from moving forward. 👍🏼 Well done. 🍻

File: modules/react/combobox/lib/ComboboxMenuList.tsx (hunk 25 above)
Comment (NicholasBoll) [reply modules/react/combobox/lib/ComboboxMenuList.tsx:22]: Typically if the props access the model, it is part of the behavior of the element. This is why we called them "behavior hooks". But some props don't access the model, but are also part of the behavior like the `role`. Some style props also are part of behavior like this orientation.

File: modules/react/menu/lib/Menu.tsx (hunk 39 above)
Comment (NicholasBoll) [reply modules/react/menu/lib/Menu.tsx:97]: Consider your life choices

File: modules/react/combobox/stories/Combobox.stories.mdx (hunk 26 above)
Comment (NicholasBoll) [reply modules/react/combobox/stories/Combobox.stories.mdx]: We updated the overall intro to combobox. Can you re-review this section?

File: modules/react/menu/lib/Menu.tsx (hunk 39 above)
Comment (NicholasBoll) [reply modules/react/menu/lib/Menu.tsx]: This section was reworked a bit. That thought was uncompleted 18 months ago. I think I meant to call out the need for text values in the collection system, but the documentation was never completed.  I've updated examples and documentation to described the purpose of text value and how to set it: https://github.com/Workday/canvas-kit/pull/2083/commits/f7d871c832668287c787226c84b465d354f2529a

Comment (NicholasBoll) [reply modules/react/menu/lib/Menu.tsx]: Nice to complete this thought a year and a half later! Thanks for catching this.

**Review by jamesfan — APPROVED** [review]
Updated intro to Combobox looks good.
